examples/custom-rlpx-subprotocol/src/main.rs

```diff
@@ -0,0 +1,298 @@
+                message: CustomRlpxProtoMessageKind::CustomMessage(msg.into()),
+            }
+        }
+
+        /// Creates a new `CustomRlpxProtoMessage` with the given message ID and payload.
+        pub fn encoded(&self) -> BytesMut {
+            let mut buf = BytesMut::new();
+            buf.put_u8(self.message_type as u8);
+            match &self.message {
+                CustomRlpxProtoMessageKind::Ping => {}
+                CustomRlpxProtoMessageKind::Pong => {}
+                CustomRlpxProtoMessageKind::CustomMessage(msg) => {
+                    buf.put(msg.as_bytes());
+                }
+            }
+            buf
+        }
+
+        /// Decodes a `CustomRlpxProtoMessage` from the given message buffer.
+        pub fn decode_message(buf: &mut &[u8]) -> Option<Self> {
+            if buf.is_empty() {
+                return None;
+            }
+            let id = buf[0];
+            buf.advance(1);
+            let message_type = match id {
+                0x00 => CustomRlpxProtoMessageId::Ping,
+                0x01 => CustomRlpxProtoMessageId::Pong,
+                0x02 => CustomRlpxProtoMessageId::CustomMessage,
+                _ => return None,
+            };
+            let message = match message_type {
+                CustomRlpxProtoMessageId::Ping => CustomRlpxProtoMessageKind::Ping,
+                CustomRlpxProtoMessageId::Pong => CustomRlpxProtoMessageKind::Pong,
+                CustomRlpxProtoMessageId::CustomMessage => {
+                    CustomRlpxProtoMessageKind::CustomMessage(
+                        String::from_utf8_lossy(&buf[..]).into_owned(),
+                    )
+                }
+            };
+            Some(Self { message_type, message })
+        }
+    }
+}
+
+/// Custom Rlpx Subprotocol Handler
+#[derive(Debug)]
+struct CustomRlpxProtoHandler {
+    state: ProtocolState,
+}
+
+impl ProtocolHandler for CustomRlpxProtoHandler {
+    type ConnectionHandler = CustomRlpxConnectionHandler;
+
+    fn on_incoming(&self, _socket_addr: SocketAddr) -> Option<Self::ConnectionHandler> {
+        Some(CustomRlpxConnectionHandler { state: self.state.clone() })
+    }
+
+    fn on_outgoing(
+        &self,
+        _socket_addr: SocketAddr,
+        _peer_id: PeerId,
+    ) -> Option<Self::ConnectionHandler> {
+        Some(CustomRlpxConnectionHandler { state: self.state.clone() })
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ProtocolState {
+    events: mpsc::UnboundedSender<ProtocolEvent>,
+}
+
+#[derive(Debug)]
+enum ProtocolEvent {
+    Established {
+        #[allow(dead_code)]
+        direction: Direction,
+        peer_id: PeerId,
+        to_connection: mpsc::UnboundedSender<Command>,
+    },
+}
+
+enum Command {
+    /// Send a custom message to the peer
+    CustomMessage {
+        msg: String,
+        /// The response will be sent to this channel.
+        response: oneshot::Sender<String>,
+    },
+}
```

Comment:
@mattsse this is cool -- we should demonstrate a bridge/oracle on this, and eventually build up to a full consensus example i feel?

and this is all subprotocols for commands post-establishing a connection, but what about discv? how should we do discovery of nodes that support custom subprotocols?